Improve the WordPress developer console with a few new features.

AllTerrainDeveloper opened the pull request:
While developing some new endpoints in WordPress.com I've realised that I needed to compare data returned from those endpoint and compare them (I was refactoring and needed to be completely sure that the endpoint was returning the same exact amount of information)

With this Pull Request I also wanted to make use of functional component, but seems that the react version we are using does not support it. So I preferred to make the change smaller and not update React at all.

This Pull Request tries to address some of the difficulties I found during that development with the WordPress developer console.

- Now it is possible to copy the JSON raw response directly with a button. Before I was opening the browser console and inspecting the network call, plus this network call had the HTML envelope and was also being a bit disturbing 
- See the response in RAW format, without the tree view.
- This PR also refactors the components a bit so they make more sense.

![image](https://user-images.githubusercontent.com/5689927/155840530-1145ff69-fb6a-4bdf-8f46-9d09f545d0d2.png)
![image](https://user-images.githubusercontent.com/5689927/155840567-888e3e34-8ee5-49a8-a20b-f93f1a2fad86.png)

### Testing instructions

Check that the default view for every call is still the Tree View. I don't want to disturb others that are used to this view.
Check that you can copy the results with the scissors button at the end of the row. You should be able to copy the contents regardless of the selected view
Check that regular calls are still working to any other endpoint you were using before (WP .COM & WP REST)

AllTerrainDeveloper commented:
I am not sure if I should fix those tests falling, seems they are testing asserts, like assert 2 == 3 for python and another for nodeJS

dmsnell commented on src/components/results/body/index.js:
```diff
@@ -0,0 +1,75 @@
+
+import './style.css';
+
+class RequestBody extends React.Component {
```
given that this is nothing more than a `render` function and that `RequestHeader` itself is just a function, is there a reason not to make `RequestBody` a single function also?

dmsnell commented on src/components/results/results-view-selector/index.js:
```diff
@@ -0,0 +1,35 @@
+	return (
+		<div className={ 'results-view-selector__button-selector-container' }>
+			<button
+				onClick={ setTreeView }
```
not sure if this gave you linter warning for creating functions inside of the `onClick` attributes, but given that this is a React function component the same thing the linter might be trying to avoid is still happening, just outside of the attributes.

we could just as easily skip some noise and declare these click handlers inline

```ts
const ResultsViewSelector = ( { … } ) => (
	<div>
		…
		<button onClick={ () => onViewChange( TREE_VIEW ) } … >Tree view</button>
		…
	</div>
);
```

dmsnell commented on src/components/results/body/index.js:
```diff
@@ -0,0 +1,75 @@
+				<JSONTree
+					theme={ jsonTheme }
+					data={ response.body }
+					shouldExpandNode={ expandFalse }
```
if we define `expandFalse` inside of `render` it will always be recreated on every render, triggering unessential re-renders. if we define it in the module scope or on the class during initialization, however, it will remain the same reference and avoid those failures to detect equality through the shallow-equality test.

dmsnell commented on src/components/results/body/index.js:
```diff
@@ -0,0 +1,75 @@
+				/>
+				}
+				{response && response.body && view === JSON_VIEW &&
+				<pre className={ 'response__json-view' } >
```
nitpick: when using static string values we don't need to jump back into the JS world, we can use the string directly

```ts
<pre className="response__json-view">
```

dmsnell commented on src/components/results/header/index.js:
```diff
@@ -1,19 +1,25 @@
 	if ( filename.indexOf( '/' ) === 0 ) {
 		filename = filename.slice( 1 );
 	}
 	filename = filename.replace( /\//g, '-' ) + '.json';
```
any reason not to do this in a single step?

```ts
const filename = path
	.replace( /^\//, '' )
	.replace( /\//g, '-' )
	+ '.json';
```

dmsnell reviewed (commented):
Nice cleanup. Seems like something missing some love and it's good getting back around to it.

dmsnell commented:
In the picture it looks like the coloration on the tree view/raw view is backwards. I would expect the emboldened or primary color to be on the active option but it looks like the active button blends in with the background.

AllTerrainDeveloper commented:
Thank you for your feedback @dmsnell! I've addressed all your concerns + added some comments in the valueRenderer function you were struggling to understand (myself too 🤣 ).

I've updated the Pull Request with the code + new screenshots showing the new colors!

dmsnell commented on src/components/results/body/index.js:
```diff
@@ -0,0 +1,74 @@
+	const customStringRenderer = value => {
+		if ( typeof value === 'string' ) {
+			const valueWithoutQuotes = value.replace( /^"|"$/g, '' );
+			let valueEscaped;
```
the way we're trying to infer types here is a bit confusing. `valueRenderer` gets multiple arguments though which can help:

```ts
const valueRenderer = (
	/** As it appears in the JSON, sometimes missing for some reason */
	valueAsString: string | null?,

	/** Parsed JSON value, with expected JS type */
	value: JSONValue,

	/** Path into original JSON value to get to here */
	...pathKey: string[]
) => { … }
```

That means we should be able to simplify this function and avoid all the awkward parsing.

```ts
const valueRenderer = ( valueAsString, value ) =>
	'string' === typeof value
		? `"${ escapeLikeJSON( value ) }"`
		: valueAsString;
```

If this is too much for this PR that's fine. Would at least be a nice follow-up.

AllTerrainDeveloper replied:
Yeah! It seems to work nicely. I didn't plan to refactor/improve that function as it was some kind of black box for me and not related to the change I meant to add, but I think it's a nice touch too.

I haven't seen any regression issue, but if you could double-check any other endpoint you might think it's conflictive it would be great, as we are not returning the same data as before because sometimes we were returning a span with some classes added to it.

dmsnell replied:
that's a good catch and the span is something I just completely overlooked. the full version with it isn't much more complicated though.

```ts
const valueRenderer = ( valueAsString, value ) => {
	// when and why does this happen?
	if ( 'string' !== typeof valueAsString ) {
		return valueAsString;
	}

	const displayValue = 'string' === typeof value
		? `"${ escapeLikeJson( value ) }"`
		: valueAsString;

	return <span className="expanded-value">{ displayValue }</span>
}
```

dmsnell commented on src/components/results/body/style.css:
```diff
@@ -0,0 +1,73 @@
+    flex-shrink: 1;
+    align-self: auto;
+    flex-grow: 1;
+    flex-shrink: 1;
```
these are duplicated from a few lines above

dmsnell commented on src/components/results/response/style.css:
```diff
@@ -0,0 +1,25 @@
+    height: 24px;
+    width: 128px;
+    border-width: 0;
+    border-width: 1px;
```
more duplicated CSS. we should remove these so that we don't confuse people in the future. "why is `border-width: 0` there? It must be there for a reason!!?"

dmsnell commented on src/components/results/header/style.css:
```diff
@@ -65,6 +65,22 @@
+}
+
+.request-header .copy:before {
+	content: "✂";
```
I noticed we're using the scissors character here but that seems suggestive of the "cut" operation instead of the "copy."

We could try using U+1F4CB "clipboard" 📋

dmsnell commented on src/components/results/response/style.css:
```diff
@@ -0,0 +1,24 @@
+    width: 128px;
+    border-width: 1px;
+    border-color: lightgrey;
+    border-style: solid;
```
if we wanted we could collapse this into one: `border: 1px solid lightgrey`

dmsnell commented on src/components/results/results-view-selector/index.js:
```diff
@@ -0,0 +1,28 @@
+
+
+const ResultsViewSelector = ( { view, onViewChange = () => {} } ) => (
+	<div className={ 'results-view-selector__button-selector-container' }>
```
no need to escape into JS and then back into a string

```ts
<div className="results-view-selector__button-selector-container">
```

dmsnell commented on src/components/results/results-view-selector/style.css:
```diff
@@ -0,0 +1,27 @@
+    display: flex;
+    border-width: 1px;
+    border-color: lightgrey;
+    border-style: solid;
```
same comment here about `border: 1px solid lightgrey`

dmsnell commented on src/components/results/index.js:
```diff
@@ -27,65 +24,15 @@ class Results extends React.Component {
+						key={ result.id }
+						result={ result }
+						className={ classnames( 'request', { error: result.response && !! result.response.error } ) }
+
```
I think there's an extra newline here we don't want

dmsnell commented on src/components/results/response/index.js:
```diff
@@ -0,0 +1,29 @@
+class Index extends React.Component {
+	componentWillMount() {
+		this.setState( { view: TREE_VIEW } );
+	}
```
do need to call this on mount? I think we should be fine setting the state to its initial value, which would eliminate a temporal dead zone where no `view` exists on the component's state.

```ts

class Index extends React.Component {
	state = {
		view: TREE_VIEW
	};

	onViewChange = view => this.setState( { view } );

	render() {
		…
	}
}
```

dmsnell commented on src/components/results/results-view-selector/index.js:
```diff
@@ -0,0 +1,28 @@
+	<div className={ 'results-view-selector__button-selector-container' }>
+		<button
+			onClick={ () => onViewChange( TREE_VIEW ) }
+			className={ `results-view-selector__button-selector ${ view === TREE_VIEW ? 'selected' : '' }` }
```
it's fine using these inline ternaries, but one thing I noticed now that I didn't before is that `classnames` is in use already in this project, even in files in this PR. so it's okay to leave it, it's okay to change it (there aren't any dependency implications or new code styles being introduced)

```ts
className={ classnames( 'results-view-selector__button-selector', {
	selected: view === TREE_VIEW
} ) }
```

dmsnell reviewed (approved):
Things look solid, thanks for being willing to go through several stages of review.

I've left some remaining comments and I want you to feel free to address them even though I gave an approval. They don't seem to have much overall impact but could clean up a few details that remain. If you do revise this again and it still tests fine there's no need to wait on me to re-review it.

AllTerrainDeveloper commented:
Sorry to ping you again @dmsnell but I am unable to deploy the changes as it fails compiling in my workstation. Any idea? I've just followed each step on the Field guide, not sure what's wrong in my environment.

```
Failed to compile.

static/js/main.bfb7c877.js from UglifyJs
SyntaxError: Unexpected token: keyword (const) [./~/jsonpointer/jsonpointer.js:55,0]
```

dmsnell commented:
Eek, looks like we have old dependencies that don't support the more modern ECMAScript features. Probably we should revert the commit and fix the build system.

We have options though:
 - Send out an immediate fix without the ES2015 features like `const`, `let`, arrow functions, etc…
 - Update our build dependencies so that they transpile them (maybe ideal?)
 - Update our build dependencies and allow them to pass-through the newer features

I'm not sure what our policy is for developer.wordpress.com given that it's public and not internal (we tend to relax our compatibility requirements for internal pages). I'm also away from my computer right now so maybe I can look into this later. I've never worked in this codebase so maybe someone else who has already would know how to update it. You might go poking around and see.

dmsnell commented:
After some initial looking I can see we have some old build configurations at play. How about we revert the change for now, update the build system, and then re-apply this patch?

AllTerrainDeveloper commented:
I've reverted it. Sorry I've been out for the whole weekend!

dmsnell commented:
No worries. Without being able to deploy we didn't break production and this repo doesn't get a lot of activity so I don't think there was any harm.

dmsnell commented:
Twist of events @AllTerrainDeveloper 

`npm run build` has been broken since `const` was introduced in #13 in _Nov 2016_.

So obviously we're not building this as indicated in the project docs. Maybe @youknowriad might remember something about how it's built. I created #88 to explore moving to `esbuild` and removing a lot of the implicit config because I couldn't figure out how to easily update `babel` and `webpack` and the `react-scripts` package without breaking everything in different ways.

dmsnell commented:
After more investigation I don't think this has been broken since 2016. I believe that an update in the `jsonpointer` library triggered a breakage at some point recently. Had we pinned all our dependencies with a `package-lock.json` we may not have noticed this because we still would have been calling for an older version of that dependency which didn't require transpiration.

Proposed fix is to transpile that dependency in #89
